RealmSwift/Impl/Persistable.swift

```diff
@@ -36,25 +36,47 @@ public protocol _Persistable: _RealmSchemaDiscoverable {
     static func _rlmGetPropertyOptional(_ obj: ObjectBase, _ key: PropertyKey) -> Self?
     // Set a value of this type on the target object
     static func _rlmSetProperty(_ obj: ObjectBase, _ key: PropertyKey, _ value: Self)
-    // Get the zero/empty/nil value for this type. Used to supply a default
-    // when the user does not declare one in their model.
-    static func _rlmDefaultValue() -> Self
     // Set the swiftAccessor for this type if the default PersistedPropertyAccessor
     // is not suitable.
     static func _rlmSetAccessor(_ prop: RLMProperty)
     // Do the values of this type need to be cached on the Persisted?
     static var _rlmRequiresCaching: Bool { get }
+    // Get the zero/empty/nil value for this type. Used to supply a default
+    // when the user does not declare one in their model. When `forceDefaultInstantiation`
+    // is true we *must* return a non-nil, default instance of `Self`. The latter is
+    // used in conjunction with key path string tracing.
+    static func _rlmDefaultValue(_ forceDefaultInstantiation: Bool) -> Self
+    // If we are in key path tracing mode, instantiate an empty object and forward
+    // the lastAccessedNames array.
+    static func _rlmKeyPathRecorder(with lastAccessedNames: NSMutableArray) -> Self
 }
+
 extension _Persistable {
     public static var _rlmRequiresCaching: Bool {
         false
     }
 }
 
-// A tag protocol for persistable types which can appear inside Optional
-public protocol _OptionalPersistable: _Persistable {
+extension _RealmSchemaDiscoverable where Self: _Persistable {
+    public static func _rlmKeyPathRecorder(with lastAccessedNames: NSMutableArray) -> Self {
+        let value = Self._rlmDefaultValue(true)
+
+        if let value = value as? ObjectBase {
+            value.lastAccessedNames = lastAccessedNames
+            value.prepareForRecording()
```

Comment:
Should return `value` here to skip trying to dynamic cast it to KeyPathStringCollection below.